### PR TITLE
support MPA router

### PR DESCRIPTION
**Pull request by TomVista**
const VueAutoRoutingPlugin = require('vue-auto-routing/lib/webpack-plugin')

    module.exports = {
    // ... other options ...

    plugins: [
       new VueAutoRoutingPlugin([
          {
         pageName:'user',
         pages: 'src/pages/user',
         importPrefix: '@/pages/user'
        },{
         pageName:'shop',
         pages: 'src/pages/shop',
         importPrefix: '@/pages/shop'
        }
        ])
      ]
    }
 get this in /node_moudle/vue-auto-routing
-user.js 
-shop.js
-index.js

and the vue-router

    import { user,shop} from 'vue-auto-routing'

    import Vue from 'vue'
    import Router from 'vue-router'

    Vue.use(Router)

    export{
       userRouter: new Router({
      mode: 'history',
      user
      }),
      shopRouter: new Router({
        mode: 'history',
        shop
      }),
    } 
the main.js  

    import { userRouter} from 'xxx'
    new Vue({
    ...
    userRouter,
    render: h => h(app),
    ...
    })

### CODE REVIEW

File: src/webpack-plugin.ts [hunk 1]
```diff
@@ -2,26 +2,43 @@ import * as assert from 'assert'
 import * as fs from 'fs'
 import * as path from 'path'
 import { Compiler } from 'webpack'
-import { generateRoutes, GenerateConfig } from 'vue-route-generator'
+import { generateRoutes } from 'vue-route-generator'
 
 const pluginName = 'VueAutoRoutingPlugin'
 
-interface Options extends GenerateConfig {}
```
Comment (ktsn) [inline src/webpack-plugin.ts:9]: Why did you remove the `extends GenerateConfig`?

File: src/webpack-plugin.ts [hunk 2]
```diff
@@ -31,10 +48,33 @@ class VueAutoRoutingPlugin {
...
+      return `${importCode}export{${exportCode}}`
+    }
 
     compiler.hooks.thisCompilation.tap(pluginName, compilation => {
       try {
-        generate()
+        if (Object.prototype.toString.call(this.options) === '[object Object]') {
+          generate(this.options as Options);
+        } else if (Array.isArray(this.options)) {
+          this.options.forEach(option => {
```
Comment (ktsn) [inline src/webpack-plugin.ts:65]: We have to validate `pageName` here as we need it if multiple options are specified.

File: README.md [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # vue-auto-routing
-
+## forked from ktsn/vue-auto-routing
```
Comment (ktsn) [inline README.md:2]: Please remove it.

File: .gitignore [hunk 4]
```diff
@@ -1,5 +1,4 @@
 node_modules/
-/lib/
```
Comment (ktsn) [inline .gitignore:2]: It shouldn't be removed.

File: lib/webpack-plugin.d.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import { Compiler } from 'webpack';
```
Comment (ktsn) [inline lib/webpack-plugin.d.ts:1]: Please revert this file as it shouldn't be committed.

File: lib/webpack-plugin.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
+"use strict";
```
Comment (ktsn) [inline lib/webpack-plugin.js:1]: Same as above, please revert this file.

File: src/webpack-plugin.ts [hunk 7]
```diff
@@ -31,10 +48,33 @@ class VueAutoRoutingPlugin {
...
+        importCode += `import ${option.pageName} from './${option.pageName}';\n`
+        exportCode += `${option.pageName},`
+      })
+      return `${importCode}export{${exportCode}}`
+    }
 
     compiler.hooks.thisCompilation.tap(pluginName, compilation => {
       try {
-        generate()
+        if (Object.prototype.toString.call(this.options) === '[object Object]') {
```
Comment (ktsn) [inline src/webpack-plugin.ts:62]:
```suggestion
        if (typeof this.options === 'object') {
```

File: src/webpack-plugin.ts [hunk 8]
```diff
@@ -31,10 +48,33 @@ class VueAutoRoutingPlugin {
...
+          })
+          const indexPath = path.resolve(__dirname, '../index.js');
+          const code = generateIndex(this.options)
+          if (fs.existsSync(indexPath) &&
+            fs.readFileSync(indexPath, 'utf8').trim() === code.trim()) {
+            return;
+          }
+          fs.writeFileSync(indexPath, code);
+        } else {
+          throw new Error('options type need object or array')
```
Comment (ktsn) [inline src/webpack-plugin.ts:76]:
```suggestion
          throw new Error('options must be an object or an array')
```

File: src/webpack-plugin.ts [hunk 9]
```diff
@@ -31,10 +48,33 @@ class VueAutoRoutingPlugin {
...
+        if (Object.prototype.toString.call(this.options) === '[object Object]') {
+          generate(this.options as Options);
+        } else if (Array.isArray(this.options)) {
+          this.options.forEach(option => {
+            generate(option)
+          })
+          const indexPath = path.resolve(__dirname, '../index.js');
+          const code = generateIndex(this.options)
+          if (fs.existsSync(indexPath) &&
+            fs.readFileSync(indexPath, 'utf8').trim() === code.trim()) {
```
Comment (ktsn) [inline src/webpack-plugin.ts:71]: We have to check every generated file including route files for each page.

File: src/webpack-plugin.ts (hunk 1 above)
Comment (TomVista) [reply src/webpack-plugin.ts:9]: fixed
```
interface Options extends GenerateConfig {
  pageName?: string;
}
```

File: src/webpack-plugin.ts (hunk 2 above)
Comment (TomVista) [reply src/webpack-plugin.ts:65]: fixed
```
} else if (Array.isArray(this.options)) {
         const isAvailable = this.options.every((value) => {
            return typeof value.pageName ==='string'
          })
          if(!isAvailable){
            throw new Error('MPA router option need pageName')
          }
```

File: .gitignore (hunk 4 above)
Comment (TomVista) [reply .gitignore:2]: fiexd

File: README.md (hunk 3 above)
Comment (TomVista) [reply README.md:2]: fixed

File: lib/webpack-plugin.d.ts (hunk 5 above)
Comment (TomVista) [reply lib/webpack-plugin.d.ts:1]: fixed

File: lib/webpack-plugin.js (hunk 6 above)
Comment (TomVista) [reply lib/webpack-plugin.js:1]: fixed

File: src/webpack-plugin.ts (hunk 7 above)
Comment (TomVista) [reply src/webpack-plugin.ts:62]: maybe
```
typeof [1, 2, 4] === 'object';

typeof new Date() === 'object';
```

File: src/webpack-plugin.ts (hunk 8 above)
Comment (TomVista) [reply src/webpack-plugin.ts:76]: fiexd

File: src/webpack-plugin.ts (hunk 9 above)
Comment (TomVista) [reply src/webpack-plugin.ts:74]: fixed

**Comment by TomVista** [comment]
for the test，
i try to do ，but I haven't written it before，so.:dog

File: package.json [hunk 10]
```diff
@@ -59,6 +59,7 @@
   "devDependencies": {
     "@types/fs-extra": "^8.0.0",
     "@types/jest": "^25.1.0",
+    "@types/mkdirp": "^1.0.1",
```
Comment (ktsn) [inline package.json:62]: Why did you add this?

File: test/__snapshots__/webpack-plugin.spec.ts.snap [hunk 11]
```diff
@@ -1264,6 +1264,108 @@ function UsersId() {
...
+/************************************************************************/
+/******/ ([
+/* 0 */
+/***/ (function(module, __webpack_exports__, __webpack_require__) {
+
+\\"use strict\\";
+__webpack_require__.r(__webpack_exports__);
+!(function webpackMissingModule() { var e = new Error(\\"Cannot find module '../../'\\"); e.code = 'MODULE_NOT_FOUND'; throw e; }());
+ // vue-auto-routing
+console.log(!(function webpackMissingModule() { var e = new Error(\\"Cannot find module '../../'\\"); e.code = 'MODULE_NOT_FOUND'; throw e; }()))
```
Comment (ktsn) [inline test/__snapshots__/webpack-plugin.spec.ts.snap:1366]: The test build doesn't seem to be succeeded.

**Review by ktsn — CHANGES_REQUESTED** [review]

**Comment by ktsn** [comment]
Sorry for the absence. I believe #149 would be better way to solve the original problem - you only need to pass multiple plugin instances with different output paths.
